chore(refactor): simplify graph ops

folded opened the pull request:
- **Simplify the implementations of graph walking.**
- **Update comments and move only_stages application**

folded committed:
Simplify the implementations of graph walking.

folded committed:
Update comments and move only_stages application

Bring only_stages closer to first/last_stages.

MattWellie commented on src/cpg_flow/workflow.py:
```diff
@@ -286,75 +304,24 @@ def _process_first_last_stages(
-            if stage.skipped:
-                graph.nodes[stage.name]['skipped'] = True
+        kept = set()
+        for node in first_stages:
+            kept.update({node} | nx.ancestors(graph, node))
+        for node in last_stages:
+            kept.update({node} | nx.descendants(graph, node))
```
Just flagging here that the graph is directional, but there's not explicit check (unless I've missed it) that the arguments to first_ and last_ are ordered - a user could accidentally swap them, there would be no validation of order, and the DAG plan could include everything after last and before first

folded replied:
There's currently a test that if if first is after last then the result is empty (which raises an error).

MattWellie commented on src/cpg_flow/workflow.py:
```diff
@@ -425,54 +391,39 @@ def set_stages(
         logger.info(f'  workflow/last_stages: {last_stages}')
 
-        # Round 1: initialising stage objects.
+        # Iinitialise stage objects.
```
```suggestion
        # Initialise stage objects.
```

MattWellie commented on src/cpg_flow/workflow.py:
```diff
@@ -163,6 +163,24 @@ def run_workflow(
 
 
+def _compute_shadow(graph: nx.DiGraph, shadow_casters: set[str]) -> set[str]:
+    """Compute the 'shadow' of a set of nodes on a directed graph.
+
+    Shadowed nodes are those that are only not reachable from any root of the
+    graph without passing through a shadow caster node."""
```
Surprised this wasn't a google pystyle [docstring](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) 👀

folded replied:
Hah. Good point. Although it's quite common for that to be enforced only on public APIs, where it matters more. The linter only complains if you look like you're attempting do document args or return types, and haven't done it properly. LMK if you want me to change it.

MattWellie commented on src/cpg_flow/workflow.py:
```diff
@@ -286,75 +304,24 @@ def _process_first_last_stages(
-                    )
-
-        if not (last_stages or first_stages):
+        if not (first_stages or last_stages):
```
I don't think it matters overly much, but this gatekeeping is already present (as in 'if not first or last, don't call this method'). This here as extra security?

folded replied:
Good point. We shouldn't have both, and I would prefer to have it in the function because that's required for its correctness. Without the early out, it would skip everything, since there are no ancestors or descendants of the empty set of nodes, so `kept` would be empty.

MattWellie commented on src/cpg_flow/workflow.py:
```diff
@@ -425,54 +391,39 @@ def set_stages(
-            ]
+        implicit_stages = set(_stages_d.keys()) - {stage.__name__ for stage in requested_stages}
+        logger.info(
+            f'Additional implicit stages: {sorted(implicit_stages)}',
```
As a UX note, it would be good to clarify the sort here. If this sort is on the String names of Stages, `sorted` is going to produce a random ordering relative to the true order, so making it less user friendly to interpret.

Looking back at the stage discovery it's probably a bit of an ordeal to order this list of Strings.

Potentially we could grow a list of encountered Stage names in `_recursively_make_stage`, flip the ordering, and use that as a start -> finish order for this logging message.

MattWellie reviewed (commented):
V interesting, looks much simpler. Shadow compute test is compelling.

folded committed:
Fix grammatical error in comment

folded committed:
Improve type info

folded committed:
remove incorrect merge conflict resolution

folded committed:
Improve type info

folded committed:
whitespace

folded committed:
disable lint: Remove quotes from type annotation

violetbrina commented:
Apologies for my delayed response.

Feel free to go ahead and bump the requests version to pass the security check.

If you could install the pre-commit hooks as well that would be great. Should have said this earlier but you can check it all out in the contirbutors file.

https://github.com/populationgenomics/cpg-flow/blob/main/CONTRIBUTING.md

If this branch contains a fix have at least one commit `fix: ...`, a new feature is `feat: ...` and for a breaking change add the exclamation after the verb so `feat!: new breaking change`

See https://www.conventionalcommits.org/en/v1.0.0/ for a full breakdown of the convention

violetbrina commented:
Just need to merge so it's not out of sync with main.

MattWellie reviewed (approved):
I like this, compelling test case, much more easily explained logical flow

rameshka reviewed (commented):
Hey @folded, thanks for taking on this work to improve the workflow logic—and sorry for taking too long to review it. 
Overall, the proposed improvements look good. I’ve left a few comments for you to take a look at. Since this logic is fairly complex, adding some inline comments would also help make it easier to follow and maintain.

MattWellie commented on src/cpg_flow/workflow.py:
```diff
@@ -237,6 +237,24 @@ def _render_node(node):
     return out
 
 
+def _compute_shadow(graph: nx.DiGraph, shadow_casters: set[str]) -> set[str]:
```
Nice breakdown, do you have the test workflow you described available to add as a test case?

I don't think your first point is a dealbreaker, `last_stages` should never contain multiple options from the same branch. Logically if a stage is included in `last_stages` it should always be run, but the option is used to define a workflow stopping point, and that should be a single stage per arm of the DAG. If we wanted to guard around this, we could add an extra variable [here](https://github.com/populationgenomics/cpg-flow/blob/dfa7e69708407784fd683434db446ff80d390725/src/cpg_flow/workflow.py#L387-L394) which keeps all Stages in first_stages and last_stages, and protects them from the later skipping action.

I can't quite get my head around `2.`. If I'm reading the DAG correctly, `A` has two successor stages, `B` & `C`, each on different branches, `D` directly follows `B`, etc. 

I'm also assuming the default run_workflow invocation contains all terminal points in the graph, so it looks like:

```
run_workflow(name='test-workflow', stages=[C, D, G])
```

* `A` will not be skipped: this is tricky, as `A` is a parent of two Stages. The first, `C`, indicates it should run through backwards discovery, as it has no first stage restriction on that branch. The other, `B` indicates it should not, which would put `A` into the pre_first set [here](https://github.com/populationgenomics/cpg-flow/blob/dfa7e69708407784fd683434db446ff80d390725/src/cpg_flow/workflow.py#L394). I would expect `A` to be inactive, as it was removed from the DAG in the 'shadow' of `B`. In practice this would leave `C` **in** the DAG, whilst the stage it continues on from is not active, not sure if that's a better outcome
* `G` will not be skipped: Not sure about this. My cold reading is that `G` would feature in `post_last` based on the `E->F->G` chain [here](https://github.com/populationgenomics/cpg-flow/blob/dfa7e69708407784fd683434db446ff80d390725/src/cpg_flow/workflow.py#L394), so it would be skipped?

Both are complications from having arbitrary workflows, and the shadow/first/last mechanism applied per branch. In practice most workflows are more linear than this, so I don't anticipate this causing problems. 

I've spent too much time thinking about this and my head hurts, but if one is available a test case to step through would be awesome

MattWellie replied:
I've added a test case corresponding to @rameshka 's description above: https://github.com/populationgenomics/cpg-flow/compare/simplify-graph-ops...simplify-graph-ops_test_case?expand=1

TL;DR Piyumi's breakdown is spot on. 

I think this is a change in logic from the OG implementation. The original & current is 
- stages are skipped if they occur after a last_stage in a contiguous branch of the DAG. 

The new implementation is 
- a stage will run unless it is after a last_stage in all possible routes through the DAG?

folded committed:
fix: restore input validation for first_stages/last_stages

The refactor dropped the check that raises WorkflowError when a stage
name in first_stages or last_stages doesn't exist; without it, nx raises
a cryptic NodeNotFound instead. Restore the validation and add tests.

folded commented:
This rebased ok. I had a look at the behaviour change you flagged with claude, and the conversation is below.

@MattWellie @rameshka What do you think? I'm not certain that this is ever an issue in practice, since I imagine that our DAGs aren't interestingly branchy. My read now is that first/last_stages is really a misfeature that encourages inconsistency in stage results.

I'm happy if the consensus is to not change the existing behaviour.

```
⏺ Rameshka identified two cases where the _compute_shadow rewrite has different semantics from the old code. MattWellie wrote a test to confirm
   it. The short version:

  The old logic was unconditional — for each first_stage, explicitly skip all its transitive dependencies; for each last_stage, explicitly skip all its transitive
  dependents. The new shadow logic is reachability-based, which means a stage can "escape" being shadowed if there's an alternative path in the DAG that reaches it without
  passing through the shadow caster.

  Concretely, for first_stages=[AB], last_stages=[AF] on:
  AA → AC
     → AB → AD
           → AE → AF → AG
           → AG
  Old result: {AA, AC, AG} skipped, {AB, AD, AE, AF} run.
  New result: AA and AG survive because AC→AA and AG has an alternative path via AB.

❯ Can you explain AA in this?
    New result: AA and AG survive because AC→AA and AG has an alternative path via AB.

  I agree that this might be a semantics change, but I'm not convinced that the old semantics were sensible.

⏺ Sure. In the new logic, _compute_shadow(graph, {AB}) asks: which nodes are unreachable from the terminal nodes without passing through AB? It walks forward from the
  leaves (AC, AD, AG) and marks nodes as "not shadowed" as it reaches them — stopping propagation only when it hits AB.

  AC is a leaf, and AC→AA is a direct edge. So AA gets reached from AC before the traversal ever touches AB. AA escapes the shadow because it's independently reachable via
  the AC branch.

  Whether that's the right behavior is a reasonable question. The old logic took a different stance: for each first_stage, unconditionally skip all its transitive
  dependencies — regardless of whether those dependencies serve other branches. So with first_stages=[AB], AA gets skipped even though AC also depends on it. AC then also
  ends up skipped because it falls outside the "kept" window. The result is the pure AB-branch only.

  The new logic's implicit reasoning is: "AA is needed by AC, which is a valid branch not blocked by any first/last constraint, so leave AA alone." That's arguably more
  coherent in a branching DAG — you avoid the situation where skipping AA orphans AC silently.

  The real question is what first_stages is supposed to mean when the first stage isn't the sole consumer of its dependencies. The old behavior effectively truncates the
  whole upstream subgraph, even nodes that other branches still need. The new behavior preserves those nodes. Neither is obviously wrong — it depends on the intended
  semantics. Do you want "skip everything before AB" (old), or "skip only what's exclusively a precursor to AB" (new)?

❯ do any of the docstrings provide an explicit description of how first/last_stages should be interpreted?

  Searched for 1 pattern, read 1 file (ctrl+o to expand)

⏺ The only docstring is the one-liner in _process_first_last_stages (lines 379–381): "Would skip all stages before first_stages, and all stages after last_stages (i.e.
  descendants and ancestors on the stages DAG.)" — and it's ambiguous in exactly the way that matters. "Before" and "after" don't say anything about what happens to
  branches that are neither before nor after, just parallel.

  There's nothing in the config documentation, run_workflow, or anywhere else that clarifies the branching case. So the intended semantics genuinely aren't written down.

❯ Here's one thing that I think is important: if we skip running AG, then its outputs are out of date wrt the rest of the graph, which seems like a dangerous result.
  Admittedly I think last_stages can always create that case. but it feels like a worse result in the old code.

  In either old or new case, can it be the case that a stage runs with a combination of old+new inputs?

⏺ Yes, it can happen in both implementations. The classic case is a diamond:

  AA → AB → AD
  AA → AC → AD

  With first_stages=[AB], AD depends on both AB (which runs fresh) and AC (which gets skipped in both old and new logic because it's not in the first→last window). AD runs
  with fresh AB outputs and stale AC outputs. Neither implementation protects against this.

  On AA specifically in the original example: you're right that the old semantics are questionable. The old logic says "AA is a dependency of AB, so skip it" — but the
  reason to skip AA is "we're assuming AB's inputs already exist." If the run includes AC, which also needs AA, then you either have to run AA (new behavior, consistent but
   defeats the checkpoint purpose) or skip AA and let AC run with stale inputs (old behavior, which is arguably the worse inconsistency since a running stage has stale
  inputs rather than a skipped stage having stale outputs).

  So to directly answer: last_stages inherently produces skipped stages whose outputs are stale relative to freshly-computed inputs — you're right that it always creates
  that case. first_stages can produce running stages that consume a mix of fresh and old inputs, which is the more dangerous variant since the outputs look valid but are
  actually based on partial recomputation. That can happen in both implementations depending on the graph shape.

  The new behavior is arguably less bad specifically for the AA case — it at least keeps the inputs to running stages consistent — but it can't solve the fundamental
  problem that partial reruns over branching graphs can't always be made coherent.
```

MattWellie commented:
My gut feelings are:

1. This is a slight change in behaviour, but it's probably fine. We don't usually have wildly branching workflows, and where we do use last_stages, it's typically on a bottleneck stage, where continuation past a certain point isn't possible in a contiguous DAG. This is usually compensating for other features of Hail, like not being able to dynamically schedule scattered workflow jobs.
2. I much prefer the new algorithm stylistically
3. I don't know if it's worth implementing this change when we're leaning towards trying to drop CPG-Flow. The current model is something we're used to, and have operated workflows using. There's a small but possible danger of current workflows misbehaving WRT this difference in function if deployed.

I think this change would be fine in practice, but the (small) risk of changed behaviour would have to be something all CPG-Flow users are aware of and are happy with. The current semantics of `first` and `last` aren't very graph oriented, but they are something we've become accustomed to.

folded commented:
Happy to drop it, then.

rameshka commented:
Hey @folded and @MattWellie, 
The differences identified come from comparing the old and new logic. As you both mentioned, the new approach likely aligns better with how our workflows behave in practice. I’m happy to go ahead with whatever you think is the best path forward after weighing pros and cons.

Thanks, @folded, for taking the time to work on these changes. 
Also, if the CPG-Flow codebase isn’t retired soon, it would be worthwhile to invest in improving test coverage + adding documentation for its maintainability.

violetbrina committed:
Merge branch 'main' into simplify-graph-ops
